File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    private List<List<Integer>> result;

    private void backtrack(int idx, int total, List<Integer> route, int[] candidates, int target) {
        if (total >= target) {
            if (total == target) {
                result.add(new ArrayList<>(route));
            }
            return;
        }

        for (int i = idx; i < candidates.length; i++) {
            if (i > idx && candidates[i] == candidates[i - 1]) {
                continue;
            }
            route.add(candidates[i]);
            total += candidates[i];
            backtrack(i + 1, total, route, candidates, target);
            total -= candidates[i];
            route.remove(route.size() - 1);
        }
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        result = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(0, 0, new ArrayList<>(), candidates, target);
        return result;
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtrack(idx, total, route):
            if total >= target:
                if total == target:
                    result.append(list(route))
                return

            for i in range(idx, len(candidates)):
                if i > idx and candidates[i] == candidates[i - 1]:
                    continue
                backtrack(i + 1, total + candidates[i], route + [candidates[i]])

        candidates.sort()
        result = list()
        backtrack(0, 0, list())
        return result